dhis2/dhis2-core
# feat: support filestore MinIO DHIS2-17174

Expose https://jclouds.apache.org/reference/javadoc/2.0.x/org/jclouds/Constants.html#PROPERTY_ENDPOINT as property `filestore.endpoint` in `dhis.conf` to support other providers like https://min.io.

We will use https://min.io in the instance manager so files/images/icons are not removed on updating an instance on play.

## Refactoring

Moved duplicated code in `JCloudsFileResourceContentStore` and `JCloudsAppStorageService` into `JCloudsStore`. Both want to read/write blobs using the jclouds `BlobStore`. The setup of `BlobStoreContext` and the root `container` for all files/images/apps is duplicated. `JCloudsStore` will now set up/tear down the `BlobStoreContext` without giving access to it. I don't think we can/should hide the JClouds API as there are too many things one can do with it that we would just create a bad abstraction. Thus, hand out a `BlobStore`, `BlobRequestSigner` or whatever short lived JClouds abstraction is needed by the `JCloudsStore` clients.

> A connection to a BlobStore in jclouds is called a BlobStoreContext. It is thread-safe and should be reused for multiple requests to the service. A BlobStoreContext associates an identity for a provider to a set of network connections.

https://jclouds.apache.org/start/blobstore/#:~:text=A%20connection%20to%20a%20BlobStore,a%20set%20of%20network%20connections.

- Move most initialization into the `JCloudsStore` constructor. We should have a valid ready to use instance once we leave the constructor. The only things that might warrant staying in `@PostConstruct` or `@PreDestroy` are creating the root container (requires disk/network access) and releasing the `BlobContext`
- Move validation that was hidden in the `BlobStoreProperties` into the constructor. Return the final value instead of using mutation.
- Rename `BlobStoreProperties` to `FileStoreConfig` as I got confused thinking this is a JClouds type. It actually only holds our filestore related properties from `dhis.conf`.
- Use parameterized log statements to avoid building unnecessary strings.
- Split credential handling from overrides so code is easier to follow without mangling unrelated things into a `Pair`.
- Put constants first in equal comparisons to avoid potential NPE.
- add common operations to the `JCloudsStore` like `getBlob`, `putBlob`, ... that are aware of us storing everything under the root `filestore.container`. Anything more involved will need to be done by getting the `BlobStore` and `container`. 

## Behavior Change

Throw an exception and thus fail to start DHIS2 if we cannot create the root container i.e. the directory or bucket named `files` by default in which we store files/icons/...

Reasons might be that the `filestore` config is wrong. This can be an invalid password in `filestore.secret`, bucket name in `filestore.container`, ...

The previous behavior was to

* [fallback](https://github.com/dhis2/dhis2-core/blob/bfa007f2e3fe2b86720820efb154726c5b37b049/dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/appmanager/JCloudsAppStorageService.java#L539) to file system store if a provider that is not supported is configured.
* [fallback](https://github.com/dhis2/dhis2-core/blob/bfa007f2e3fe2b86720820efb154726c5b37b049/dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/appmanager/JCloudsAppStorageService.java#L543) to an in-memory (`transient`) store if file system store is configured but `!locationManager.externalDirectorySet()`
* [catch](https://github.com/dhis2/dhis2-core/blob/bfa007f2e3fe2b86720820efb154726c5b37b049/dhis-2/dhis-services/dhis-service-core/src/main/java/org/hisp/dhis/appmanager/JCloudsAppStorageService.java#L181) `AuthorizationException` and log an error. This cannot be recovered from. Any code relying on the store will fail.

The previous behavior neglects what the user intended. In some cases that leads to more work or even data loss (picking provider `transient`). Logging and starting up hides the root cause.

* Remove the validation of `filestore.container` via

```java
  private static final Pattern CONTAINER_NAME_PATTERN =
      Pattern.compile("^(?![.-])(?=.{1,63})([.-]?[a-zA-Z0-9]+)+$");
```

We seemed to try to validate according to https://docs.aws.amazon.com/AmazonS3/latest/userguide/bucketnamingrules.html

The pattern did not capture the S3 rules correctly as a name should be 63 (max) characters long.

```java
  @Test
  void test() {
    String name = "aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa";
    assertTrue(name.length() > 63); // which makes the name an invalid bucket name

    assertFalse(isValidContainerName(name)); // evaluates to true
  }
```

Providers might follow different rules. We would need correctly capture all rules of all providers. Using one validation would mean creating the most strict one. This means we also needlessly restrict anyone for example when using the file system. We removed this validation as we now fail on startup if we cannot create the root container. The validation is thus done by the provider only.

## Test

Added
* a unit test for `JCloudsStore` throwing if unsupported provider is configured, external dir is not set on file system provider.
* an e2e test that configures DHIS2 to store file resources in minio.

Manual tests
* upload a birth certificate via the capture app and download it again
* install an older version of Capture app via App Management
```
tooling-web-1            | INFO  org.hisp.dhis.appmanager.JCloudsAppStorageService [http-nio-8080-exec-9] Installing new app: capture_100.67.8.zip
tooling-web-1            | INFO  org.hisp.dhis.appmanager.JCloudsAppStorageService [http-nio-8080-exec-9] Starting JClouds discovery
tooling-web-1            | INFO  org.hisp.dhis.appmanager.JCloudsAppStorageService [http-nio-8080-exec-9] Found potential app: apps/capture_100.67.8/
tooling-web-1            | INFO  org.hisp.dhis.appmanager.JCloudsAppStorageService [http-nio-8080-exec-9] New app 'Capture' installed
tooling-web-1            |      Install path: apps/capture_100.67.8
```

* Note: we cannot delete/install yet another version of Capture app via App Management see https://dhis2.atlassian.net/browse/DHIS2-17197
